stock_storage_type/models/stock_location.py

```diff
@@ -141,21 +141,48 @@ def _compute_leaf_location_ids(self):
             leaves = self.search([("id", "in", leave_ids)])
             loc.leaf_location_ids = leaves
 
-    @api.depends("quant_ids", "in_move_ids", "in_move_line_ids")
+    def _should_compute_will_contain_product_ids(self):
+        return any(
+            location.do_not_mix_products
+            for location in self.allowed_location_storage_type_ids
+        )
+
+    def _should_compute_will_contain_lot_ids(self):
+        return any(
+            location.do_not_mix_lots
+            for location in self.allowed_location_storage_type_ids
+        )
+
+    @api.depends(
+        "quant_ids",
+        "in_move_ids",
+        "in_move_line_ids",
+        "allowed_location_storage_type_ids.do_not_mix_products",
+    )
     def _compute_location_will_contain_product_ids(self):
         for rec in self:
-            rec.location_will_contain_product_ids = (
-                rec.mapped("quant_ids.product_id")
-                | rec.mapped("in_move_ids.product_id")
-                | rec.mapped("in_move_line_ids.product_id")
-            )
+            products = self.env["product.product"].browse()
```

Comment:
do you expect to have many records here? If yes, `set` of ids would be better... as you know :wink: 
Same below.

Reply:
We have 3 sets of products to union (quants, incoming moves, incoming move lines), what's slow is iterating thousands of times to union record id(s).